common/rest_types/src/handler.rs

```diff
@@ -0,0 +1,247 @@
+                    e
+                ))
+            })
+            .map(String::from)?;
+
+        Ok(Self {
+            executor,
+            req,
+            body,
+            ctx,
+            allow_body: false,
+            encoding: ApiEncodingFormat::from(accept_header.as_str()),
+        })
+    }
+
+    /// The default behaviour is to return an error if any body is supplied in the request. Calling
+    /// this function disables that error.
+    pub fn allow_body(mut self) -> Self {
+        self.allow_body = true;
+        self
+    }
+
+    /// Return a simple static value.
+    ///
+    /// Does not use the blocking executor.
+    pub async fn static_value<V>(self, value: V) -> Result<HandledRequest<V>, ApiError> {
+        // Always check and disallow a body for a static value.
+        let _ = Self::get_body(self.body, false).await?;
+
+        Ok(HandledRequest {
+            value,
+            encoding: self.encoding,
+        })
+    }
+
+    /// Calls `func` in-line, on the core executor.
+    ///
+    /// This should only be used for very fast tasks.
+    pub async fn in_core_task<F, V>(self, func: F) -> Result<HandledRequest<V>, ApiError>
+    where
+        V: Send + Sync + 'static,
+        F: Fn(Request<Vec<u8>>, T) -> Result<V, ApiError> + Send + Sync + 'static,
+    {
+        let body = Self::get_body(self.body, self.allow_body).await?;
+        let (req_parts, _) = self.req.into_parts();
+        let req = Request::from_parts(req_parts, body);
+
+        let value = func(req, self.ctx)?;
+
+        Ok(HandledRequest {
+            value,
+            encoding: self.encoding,
+        })
+    }
+
+    /// Spawns `func` on the blocking executor.
+    ///
+    /// This method is suitable for handling long-running or intensive tasks.
+    pub async fn in_blocking_task<F, V>(self, func: F) -> Result<HandledRequest<V>, ApiError>
+    where
+        V: Send + Sync + 'static,
+        F: Fn(Request<Vec<u8>>, T) -> Result<V, ApiError> + Send + Sync + 'static,
+    {
+        let ctx = self.ctx;
+        let body = Self::get_body(self.body, self.allow_body).await?;
+        let (req_parts, _) = self.req.into_parts();
+        let req = Request::from_parts(req_parts, body);
+
+        let value = self
+            .executor
+            .clone()
+            .handle
+            .spawn_blocking(move || func(req, ctx))
+            .await
+            .map_err(|e| {
+                ApiError::ServerError(format!(
+                    "Failed to get blocking join handle: {}",
+                    e.to_string()
+                ))
+            })??;
+
+        Ok(HandledRequest {
+            value,
+            encoding: self.encoding,
+        })
+    }
+
+    /// Call `func`, then return a response that is suitable for an SSE stream.
+    pub async fn sse_stream<F>(self, func: F) -> ApiResult
+    where
+        F: Fn(Request<()>, T) -> Result<Body, ApiError>,
+    {
+        let body = func(self.req, self.ctx)?;
```

Comment:
This callback style is hiding away the OS thread creation.  The plan is to have a [single OS thread shared among several (SSE) HTTP endpoints](https://github.com/sigp/lighthouse/pull/1107#discussion_r420665464), so I think it should be brought into forefront.  Also, instead of passing around requests/responses as function arguments/return values, I'd suggest passing around synchronization primitives (channel, bus).  Requests/responses do not reflect the actual program control flow in case of hyper's `Body::channel()`.